### PR TITLE
Build improvements and optimizations

### PR DESCRIPTION
**Pull request by madison-ookla**
This PR includes a series of improvements to the build infrastructure within this custom setup. There are multiple motivations for this:
* Improving build time for python-only config changes, allowing for more rapid iteration when developing
* Reducing image size (my machine kept running out of disk space while building the images!)

### Changes
* Added `set -e` to the `static_build.sh` script so any error on any command fails the build without needing the extra validation function.
* Moved the static components out of the `config` folder for the frontend image. This means that changes to the python `config.py` or `requirements.txt` no longer require recompiling all static components.
* Reduced image size by removing unnecessary `usermod`/`chown` steps (due to Docker's layer's image file system, this was effectively doubling image size and I was not able to determine why it was necessary).
* Reduced image size by combining multiple steps where possible.
* For all image, re-ordered steps so changes to `config.py` did not require reinstalling all python requirements.
* Used `CMD` exclusively so that alternate run commands could be handed to the containers (e.g. `docker-compose run amundsenfrontend ls`, previously this would fail because `gunicorn` was defined as the entrypoint and would not understand the argument `ls`).
* Reduced the image size of the metadata service by using the `slim` python 3.7 image.
* Moved the dockerfiles into the respective folders. This has the twofold benefit of 1) not having to define the `dockerfile` in the `docker-compose.yml` and 2) reducing the size of the build context that's sent for each service to the Docker daemon, as only the service-specific folders are needed for the build anyway.

Let me know if any of that needs more explanation! I've been running this setup locally with success and it has made rapid iteration of the python config files much easier.

### CODE REVIEW

**Comment by verdan** [comment]
Thank you so much @madison-ookla for refactoring this. I will make sure to verify this implementation this week. 💯

**Review by sewardgw — APPROVED** [review]
Tested this as well, running a bunch of data through the built containers. Build times are anywhere from 30%-70% faster, depending on the changes made.

We should probably look to automate tests within a docker container at some point to ensure there are no build dependency errors that occur based on the image being retrieved.

File: docker-compose.yml [hunk 1]
```diff
@@ -32,8 +32,7 @@ services:
         - ./.local/elasticsearch/data:/usr/share/elasticsearch/data
   amundsensearch:
       build:
-        context: .
-        dockerfile: Dockerfile.search
+        context: search
```
[the remark this answers is not in the document]
Comment (madison-ookla) [reply docker-compose.yml:35]: Exactly! And we don't have to specify the dockerfile explicitly now either 😄

File: search/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
+FROM python:3.7-slim
```
[the remark this answers is not in the document]
Comment (madison-ookla) [reply search/Dockerfile:1]: Moving from slim to alpine removes a lot of basic tooling that python packages frequently need (`gcc`, other system libs, etc). I've often found that it's easier to just use the higher image size that comes with slim than try and manage installing every system dependency you need on alpine. We've found too with some other images that, when we do have to install more system libraries for python packages, the size gets pretty close to slim anyway 🤷🏼‍♀️

**Comment by madison-ookla** [comment]
Any chance this can get merged? 😄

**Review by hhobson — APPROVED** [review]
Nice improvements @madison-ookla . Have implemented them, so thanks. 

In recent versions of Docker only `RUN`, `ADD` & `COPY` generate layers and other commands don't impact size. So not sure if combining all `ENV`s into single step will have contributed to reduction in size, but didn't test this when I updated my images. 

Would be good to know what the intention/use of the `amundsen` user was/is.

**Comment by madison-ookla** [comment]
> In recent versions of Docker only `RUN`, `ADD` & `COPY` generate layers and other commands don't impact size. So not sure if combining all `ENV`s into single step will have contributed to reduction in size, but didn't test this when I updated my images.

Woah, this is news to me! Do you have any links to documentation that discusses this that you could share?

> Would be good to know what the intention/use of the `amundsen` user was/is.

Typically, it's a good idea _not_ to run as the root user within a docker container in production. However the user steps necessary here were significantly bloating image/layer size, so I opted to remove them.

**Comment by hhobson** [comment]
Sorry for delayed response. 

> Woah, this is news to me! Do you have any links to documentation that discusses this that you could share?

Change to layer generating commands is mentioned in the [docs Here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers). I only came across it by mistake.

Thanks for the heads up on the User. Read into it and seems quite important. What I've done is added a `RUN` statement before the first `COPY` that calls `groupadd` & `useradd`, then included a `--chown=amundsen:amundsen` flag in `COPY` commands. Also explicit set the user with `USER` command. Can share with you if you're interested.

**Comment by hhobson** [comment]
Can use `.dockerignore` to exclude redundant upstream files. Believe you can ignore everything but the service directory, `MANIFEST.in`, `requirements.txt`, `setup.cfg` & `setup.py` files.

**Pull request merged**